Support AKP153E based on AKP153

teras opened the pull request:
This PR supports AKP153E based on AKP153 code. Actually it mirrors everything except the IDs and [the observation](https://github.com/streamduck-org/elgato-streamdeck/issues/13#issuecomment-2267174135) made by @d-kja

Discussion about this patch also [here](https://github.com/ninjadev64/OpenDeck/issues/10).

ninjadev64 commented:
I would rename the `AJAZZ_E_VENDOR_ID` constant, as someone with an Ajazz AKP153**R** reported the same vendor ID today (the PID is 0x1020 however).

teras commented:
Yes he [already](https://github.com/streamduck-org/elgato-streamdeck/issues/13#issuecomment-2407761321) asked for support. Any proposals how to name it?  `AJAZZ2_VENDOR_ID` ? `AJAZZX_VENDOR_ID` ?

ninjadev64 commented:
`AJAZZ_VENDOR_ID_1` and `AJAZZ_VENDOR_ID_2` probably

teras commented:
> `AJAZZ_VENDOR_ID_1` and `AJAZZ_VENDOR_ID_2` probably

done

ninjadev64 commented:
Minor, but there is also a missing "P" in "AKP" in most of the comments (from ZCube's original contribution)

teras commented:
You are right, I didn't notice it when I was performing copy & paste :rofl: 
Now it should be OK

ninjadev64 commented:
The comments here are also wrong unfortunately
```
/// Vendor ID of Ajazz Stream Deck AKP153E
pub const AJAZZ_VENDOR_ID_2: u16 = 0x0300;

/// Product ID of Ajazz Stream Deck AKP153
pub const PID_AJAZZ_E_AKP153E: u16 = 0x1010;
```

teras commented:
Only the bottom one, right?

teras commented:
Since right now this driver won't support the R version, I believe it's misleading to add it in the comment, in the same way that we don't know exactly how many (if any) other devices with the original vendor id exist. I'd propose it to keep it this way and if/when the driver is able to support more, then the commend could be updated.

Moreover I updated the udev rules (which ironically I had updated the file in my own system :smile: ) and fixed the ACK rule to properly support it without breaking the buttons

teras commented:
It's **not ready** to be merged. There's [a bug in the patch of @ZCube](https://github.com/streamduck-org/elgato-streamdeck/issues/15) that doesn't allow proper parsing in async mode. I have [the same problem](https://github.com/ninjadev64/OpenDeck/issues/26) too.

TheJebForge commented:
Can you draft the pull request until then?

teras commented:
I am trying to find ways, without destroying what's there, to bypass the problem. Should the PR close? I think we're missing the discussions then.

TheJebForge commented:
There was a small link on top right to convert this PR to a draft, not sure if you had it, but probably did

ninjadev64 commented on examples/async/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "simple"
```
This should probably be changed.

TheJebForge commented:
By the way, why you keep force pushing? You can keep adding new commits and I'll just squash merge it later

teras commented:
Ermmm most people demand to force push instead of adding more commits. I was even rejected once because I didn't force push :laughing: 

Creating more commits is more easy of course.

Now, I managed to make it work and fix also the previous AKP153 commit, together with my AKP153E.
I'd say it is completed except if you want for me to make any other changes.

TheJebForge commented:
They're not aware of "Squash and merge" apparently...

teras commented:
You're actually angels compared to them

TheJebForge commented:
Imo looks good. I will do a change later to list_devices since if we'll keep getting more vendor IDs to track, might as well wrap that into some sort of enum. But have yet to decide how that's gonna look like..

ninjadev64 commented:
So I presume projects utilising this crate will have to add device.initialise() when they update?

teras commented:
Yes @ninjadev64 

Also super important: you always need to flush the images when you send them to the device. You don't have to flush them one by one. You can batch send all images and flush at the end. If no image is send at all, no flushing is required; that's why there's the original check 
```rust
if device.is_updated() {
    device.flush().unwrap();
}

```

If an image is send but no flushing is performed, the device freezes. Right now OpenDeck doesn't work because ( I assume ) no flushing is performed (in the asynchronous version of the API there were no flush method).

ninjadev64 commented:
Alright. I presume the flushing is only needed for Ajazz devices as it works just fine with Elgato ones. I think the library's API should abstract over the different hardware vendors better so calling initialise and flush is not left out by someone developing against an Elgato device rendering Ajazz ones non-functional.

@TheJebForge thoughts?

teras commented:
It already does. Just flush and it's not your job to worry which needs flushing and which does not. That's what abstractions are for.

The library had already hooks for other devices (even doing nothing but just to update an atomic boolean); same goes for initialization.

TheJebForge commented:
Maybe flushing should be implemented for Elgato devices where write operations will be buffered until you call flush, so users of this library won't forget to flush for Ajazz devices where it's actually required. Initialization could be part of connection instead of being a separate method you have to call?

teras commented:
For flushing, I was tempted to actually flush every time an image is sent. But this is not the "correct" way to do it, especially in embedded devices.

I'd say that, everyone should flush, no matter what, even if no needed. It's like flushing on other I/O ports, even if sometimes the communication is synchronous and not required. For example, flushing on a memory block is not required, but  the API doesn't change just because it's not. You still need to flush and close the I/O stream.

About initializing, that's indeed debatable if it should be in the connection method or not. IMHO these are two separated things and shouldn't be mixed. Hiding something irrelevant just because it might make the API cleaner, I am not sure how "clean" the API itself would be. Yes, it's a ritual, yes it's one more step, but the procedure is complicated after all and this is a library; the person who is going to use it should follow the steps fanatically anyway.

TheJebForge commented:
Unfortunately, some people don't read manuals, and if from experimentation it seems that for example, just writing the image works on their device, they won't even question why should they flush. They see that it works for them and will continue to do it like that, if flushing would be required for all of the devices, they would see how it doesn't work unless they flush, and that would achieve full support for Ajazz devices too

TheJebForge commented:
Actually, what if there was a separate wrapper for the device type where it would actually handle all of the things automatically for simple use cases, so both simple and advanced users of the library would be satisfied

ninjadev64 commented:
> Unfortunately, some people don't read manuals, and if from experimentation it seems that for example, just writing the image works on their device, they won't even question why should they flush. They see that it works for them and will continue to do it like that, if flushing would be required for all of the devices, they would see how it doesn't work unless they flush, and that would achieve full support for Ajazz devices too

My thoughts exactly

teras commented:
Having two wrappers for the device I believe it makes things worse, since it will confuse which one to use and they will always use the wrong one. It's a law of nature :)

If you believe that the initialization should be under the hood, it's your decision.

TheJebForge commented:
Ehh, that confusion can be worked around with design, will see..

teras commented:
I had an idea: what if we add an atomic boolean under the hood  to mark the call to a lazy initialization, the second we start writing to this device. Then we don't add any new methods and the device gets initialized the minute it is required.

About the flush image I tend to agree with @TheJebForge . When there's a request to update an image, instead of directly sending the image, keep it in memory until `flush()` is performed. Thus all devices will have exactly the same behavior and the devices that need extra steps, like the Ajazz ones, will be supported too. Actually this might be a good idea anyway, adding practically zero overhead on the execution.

What do you think? If you agree I can start working on it.

ninjadev64 commented:
Both of those sound good to me, but of course I am not the developer of this library so wait for @TheJebForge's response.

TheJebForge commented:
Please wrap image_cache into some inner mutability, setting any of the methods to be `&mut self` stops you from being able to do anything once you get a reader for a device

teras commented:
> Please wrap image_cache into some inner mutability, setting any of the methods to be `&mut self` stops you from being able to do anything once you get a reader for a device

done. Sorry :)

TheJebForge commented:
Sorry for delay, I get too tired at work to do anything on the side. I think I'll merge this into a separate branch first and do some of my own changes. Initialization should probably be part of connecting to the device, since you would connect to the device when you actually want to start using it... Once am done with my changes, will push out a new major version of the library (0.8 since API got changed somewhat)

TheJebForge commented:
Nvm, you already had initialize as private method whenever anything is happening

ninjadev64 commented:
Awesome, I'll update OpenDeck soon.
